clang/test/C/C23/n3037.c

```diff
@@ -0,0 +1,472 @@
+                         c23-note {{attribute 'deprecated' here}}
+};
+
+enum enum_attr_test_5 { // c17-error {{redefinition of 'enum_attr_test_5'}} \
+                           c23-error {{type 'enum enum_attr_test_5' has incompatible definition}}
+  EAT5 // c17-error {{redefinition of enumerator 'EAT5'}} \
+          c23-note {{no corresponding attribute here}}
+};
+
+enum enum_attr_test_6 { // c17-note {{previous definition is here}}
+  EAT6 // c17-note {{previous definition is here}} \
+          c23-note {{no corresponding attribute here}}
+};
+
+enum enum_attr_test_6 { // c17-error {{redefinition of 'enum_attr_test_6'}} \
+                           c23-error {{type 'enum enum_attr_test_6' has incompatible definition}}
+  EAT6 [[deprecated]] // c17-error {{redefinition of enumerator 'EAT6'}} \
+                         c23-note {{attribute 'deprecated' here}}
+};
+
+// You cannot declare one with a fixed underlying type and the other without a
+// fixed underlying type, or a different underlying type. However, it's worth
+// showing that the underlying type doesn't change the redefinition behavior.
+enum fixed_test_1 : int { FT1 }; // c17-note 2 {{previous definition is here}}
+enum fixed_test_1 : int { FT1 }; // c17-error {{redefinition of 'fixed_test_1'}} \
+                                    c17-error {{redefinition of enumerator 'FT1'}}
+
+enum fixed_test_2 : int { FT2 };                 // c17-note 2 {{previous definition is here}}
+enum fixed_test_2 : typedef_of_type_int { FT2 }; // c17-error {{redefinition of 'fixed_test_2'}} \
+                                                    c17-error {{redefinition of enumerator 'FT2'}}
+
+// Test more bizarre situations in terms of where the type is declared. This
+// has always been allowed.
+struct declared_funny_1 { int x; }
+declared_funny_func(struct declared_funny_1 { int x; } arg) { // both-warning {{declaration of 'struct declared_funny_1' will not be visible outside of this function}}
+  return declared_funny_func((__typeof__(arg)){ 0 });
+}
+
+// However, this is new.
+struct Outer {
+  struct Inner { // c17-note {{previous definition is here}}
+    int x;
+  } i;
+
+  enum InnerEnum { // c17-note {{previous definition is here}}
+    IE1            // c17-note {{previous definition is here}}
+  } j;
+};
+
+struct Inner {   // c17-error {{redefinition of 'Inner'}}
+  int x;
+};
+
+enum InnerEnum { // c17-error {{redefinition of 'InnerEnum'}}
+  IE1            // c17-error {{redefinition of enumerator 'IE1'}}
+};
+
+void hidden(void) {
+  struct hidden_struct { int x; };
+}
+
+struct hidden_struct { // This is fine because the previous declaration is not visible.
+  int y;
+  int z;
+};
+
+struct array { int y; int x[]; };    // c17-note {{previous definition is here}} \
+                                        c23-note {{field 'x' has type 'int[]' here}}
+struct array { int y; int x[0]; };   // c17-error {{redefinition of 'array'}} \
```

Comment:
I just wondered, what about VLAs?

Reply:
Can't have a VLA in a structure, so there's nothing there to test.